### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by klzns**
This app was created to ease the read and manipulation of session fields.

Inspired by [this thread on Slack](https://vtex.slack.com/archives/CDQSTD7N1/p1611151477004300).

You can read the rendered documentation here:
https://github.com/vtex-apps/session-client/tree/feature/initial-implementation

Example of usage of the components examples of the documentation:
https://session--storeframework.myvtex.com/

You can use this branch of the theme to test: https://github.com/vtex/storeframework.store-theme/tree/session-client

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    }
+    ... on SessionError {
+      type
+      message
+    }
+  }
+}
+```
+
+#### updateSesion Mutation
```
Comment (icazevedo) [inline docs/README.md]: Just a typo

```suggestion
#### updateSession Mutation
```

File: manifest.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "session-client",
+  "vendor": "vtex",
+  "version": "0.2.1",
```
Comment (icazevedo) [inline manifest.json:4]: Why 0.2.1? 🤔

File: node/package.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+  "dependencies": {
+    "cookie": "^0.4.1",
+    "ramda": "^0.26.1"
```
Comment (icazevedo) [inline node/package.json]: I don't think you need that :harold-pain:

File: node/resolvers/session.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+export async function session(
+  _: any,
+  { items }: Args,
+  context: ServiceContext<Clients>
+) {
+  const {
+    clients: { customSession },
+  } = context
+
+  const sessionCookie = getSessionToken(context.cookies)
```
Comment (icazevedo) [inline node/resolvers/session.ts:21]: Maybe add cookies to the context  destructuring

File: node/resolvers/updateSession.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+export async function updateSession(
+  _: any,
+  { fields, items }: Args,
+  context: ServiceContext<Clients>
+) {
+  const {
+    clients: { customSession },
+  } = context
+
+  const sessionCookie = getSessionToken(context.cookies)
```
Comment (icazevedo) [inline node/resolvers/updateSession.ts:20]: Same here

File: react/useUpdateSession.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    options
+  )
+
+  const updateSession = useCallback(
+    async (mutationOptions: MutationFunctionOptions<Data, Variables>) => {
+      await updateSessionMutation(mutationOptions)
+      window.location.reload()
+
+      // Holds promise until page reload
+      return new Promise(() => {})
```
Comment (icazevedo) [inline react/useUpdateSession.tsx:32]: Really good

**Review by icazevedo — CHANGES_REQUESTED** [review]
Awesome work! The docs are really good

File: docs/README.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+- [`useLazyFullSession`](#useLazyFullSession)
+
+To update session:
+
+- [`useUpdateSession`](#useUpdateSession)
+- [`useUpdateSessionInline`](#useUpdateSessionInline)
+
+#### `useRenderSession`
+
+The fastest response to a session value. It uses the session response from render-session. One caveat: the session values are limited to a [set of values](https://github.com/vtex-apps/render-session/blob/master/src/constants.ts). If you need fields that are not in this set, you can use [`useFullSession`](#useFullSession) or [`useLazyFullSession`](#useLazyFullSession).
```
Comment (victorhmp) [inline docs/README.md]:
```suggestion
The fastest way to access a session value. It uses the session response from render-session. One caveat: the session values are limited to a [set of values](https://github.com/vtex-apps/render-session/blob/master/src/constants.ts). If you need fields that are not in this set, you can use [`useFullSession`](#useFullSession) or [`useLazyFullSession`](#useLazyFullSession).
```

File: docs/README.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  }
+
+  console.log({ session: data?.session })
+
+  return <>Session is ready</>
+}
+
+export default MyComponent
+```
+
```
Comment (victorhmp) [inline docs/README.md:93]:
```suggestion
```

File: docs/README.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+It also accepts a GraphQL variable called `items` which is an array of strings of session items to fetch only certain items of the session.
```
Comment (victorhmp) [inline docs/README.md]:
```suggestion
It also accepts a GraphQL variable called `items` which is an array of `string`s. These `string`s should match attributes inside of the Session object, and only those attributes will be fetched and returned by `useFullSession`.
```

File: docs/README.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+useFullSession({
+  variables: {
+    items: ['store.channel', 'store.countryCode']
+  }
+})
+```
+
+#### `useLazyFullSession`
+
+The same as [`useFullSession`](#useFullSession) but it uses the React Apollo's `useLazyQuery` hook. You can [read more about `useLazyQuery` API](https://www.apollographql.com/docs/react/api/react/hooks/#uselazyquery).
```
Comment (victorhmp) [inline docs/README.md]:
```suggestion
The same as [`useFullSession`](#useFullSession) but it uses React Apollo's `useLazyQuery` hook. You can [read more about `useLazyQuery` API](https://www.apollographql.com/docs/react/api/react/hooks/#uselazyquery).
```

File: node/resolvers/session.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import { ServiceContext } from '@vtex/api'
+
+import { Clients } from '../clients'
+import { getSessionToken } from '../modules/sessionToken'
+
+export const VTEX_SESSION = 'vtex_session'
+
+interface Args {
+  items?: string[]
```
Comment (victorhmp) [inline node/resolvers/session.ts:9]: Do we know all valid strings?

File: react/graphql/session.graphql [hunk 12]
```diff
@@ -0,0 +1,12 @@
+query session($items: [String]) {
+  session(items: $items) @context(provider: "vtex.session-client") {
+    ... on SessionSuccess {
```
Comment (victorhmp) [inline react/graphql/session.graphql:3]: Wow, I've never seen this syntax 🤯

File: react/useFullSession.tsx [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+import type { Session } from './SessionTypes'
+import SessionQuery from './graphql/session.graphql'
+
+interface Data {
+  session: Session
+}
+
+interface Variables {
+  items?: string[]
```
Comment (victorhmp) [inline react/useFullSession.tsx:11]: Same here, can we make this safer somehow?

**Review by victorhmp — CHANGES_REQUESTED** [review]
Awesome! We should update `add-to-cart-button` to use this!

File: node/resolvers/session.ts (hunk 11 above)
Comment (klzns) [reply node/resolvers/session.ts:9]: A priori, no.

File: react/graphql/session.graphql (hunk 12 above)
Comment (klzns) [reply react/graphql/session.graphql:3]: ~~ UNION TYPES 🔫  pew pew

File: react/useFullSession.tsx (hunk 13 above)
Comment (klzns) [reply react/useFullSession.tsx:11]: It can be anything

File: manifest.json (hunk 2 above)
Comment (klzns) [reply manifest.json:4]: I started from `graphql-example` app which had this version 🤷  haha

**Review by icazevedo — APPROVED** [review]

**Review by victorhmp — APPROVED** [review]

File: .vscode/extensions.json [hunk 14]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (igorbrasileiro) [inline .vscode/extensions.json:1]: This file should be added?! If not, add `.vscode` to gitignore

File: .vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (igorbrasileiro) [inline .vscode/settings.json:1]: This file should be added?! If not, add .vscode to gitignore

File: .vscode/extensions.json (hunk 14 above)
Comment (victorhmp) [reply .vscode/extensions.json:1]: Yep, it's part of our template

File: .vscode/settings.json (hunk 15 above)
Comment (victorhmp) [reply .vscode/settings.json:1]: Also part of our template

**Pull request merged**